### PR TITLE
[202405][Rebase&&FF] Firmware Performance

### PR DESCRIPTION
**Pull request by Flickdm**
This PR adds a series of commits intended to add firmware performance measurements to a platform.

### Overview of commits:

1. MdePkg: Add EFI_PRE_READY_TO_BOOT_GUID and EFI_POST_READY_TO_BOOT_GUID
   1. This commit adds two new GUIDS required for Delay Dispatch
2. REBASE: Revert "MdePkg: Added header file for Delayed Dispatch PPI"
   1. This commit reverts a commit brought in from upstream that modifies the DelayDispatch.h file in MdePkg that was moved to MdeModulePkg
 3. MdeModulePkg/MdeModulePkg.dec: Adds Delay Dispatch
    1. This commit adds Delay Dispatch
 4. MdeModulePkg: Adds Performance Measurement V2
    1. This commit adds the initial performance measurements
 5. Perf-instrument PreReadyToBoot, ReadyToBoot, PostReadyToBoot signals …
    1. This commit adds the signals required for performance measurements
 6. Merged PR 5181: Change FirmwarePerformanceDxe.c EndofDxeEvent tpl to …
    1. This commit changes the TPL_NOTIFYs to TPL_CALLBACKS
 7. MdeModulePkg: PeiMain: Adds performance measurement for PeiMain
    1. Adds peformance measurements for PEI

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

----
### How This Was Tested
From 2311.

### CODE REVIEW

File: MdeModulePkg/Core/Dxe/Dispatcher/Dispatcher.c [hunk 1]
```diff
@@ -414,6 +414,7 @@ CoreDispatcher (
     //
     // If the dispatcher is running don't let it be restarted.
     //
+    PERF_FUNCTION_END (); // MS_CHANGE
```
Comment (makubacki) [inline MdeModulePkg/Core/Dxe/Dispatcher/Dispatcher.c]: Please try to replace these with `MU_CHANGE` if you can before publishing the PR.

File: MdePkg/Include/Guid/EventGroup.h [hunk 2]
```diff
@@ -49,4 +49,15 @@ extern EFI_GUID  gEfiEventDxeDispatchGuid;
 
 extern EFI_GUID  gEfiEndOfDxeEventGroupGuid;
 
+// MU_CHANGE [BEGIN]
+#define EFI_PRE_READY_TO_BOOT_GUID \
```
Comment (os-d) [inline MdePkg/Include/Guid/EventGroup.h:54]: Can you split this into two commits, first for MdePkg, second for MdeModulePkg?

**Review by Javagedes — APPROVED** [review]

File: MdePkg/Include/Ppi/DelayedDispatch.h [hunk 3]
```diff
@@ -1,102 +0,0 @@
-/** @file
```
Comment (os-d) [inline MdePkg/Include/Ppi/DelayedDispatch.h]: I assume you need this commit to build for changes? Seems unrelated otherwise.

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 4]
```diff
@@ -5,6 +5,8 @@ Copyright (c) 2006 - 2024, Intel Corporation. All rights reserved.<BR>
 (C) Copyright 2016 Hewlett Packard Enterprise Development LP<BR>
 SPDX-License-Identifier: BSD-2-Clause-Patent
 
+Copyright (c) 2017, Microsoft Corporation.
```
Comment (os-d) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: Copyright (C) Microsoft Corporation.

Please ensure any additions have this format.

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 5]
```diff
@@ -1411,6 +1413,24 @@ PeiDispatcher (
...
+      UINT32  TableSize = sizeof (DELAYED_DISPATCH_TABLE) + ((FixedPcdGet32 (PcdDelayedDispatchMaxEntries) - 1) * sizeof (DELAYED_DISPATCH_ENTRY));
+      Private->DelayedDispatchTable = BuildGuidHob (&gEfiDelayedDispatchTableGuid, TableSize);
+      if (NULL != Private->DelayedDispatchTable) {
+        ZeroMem (Private->DelayedDispatchTable, TableSize);
+        Status = PeiServicesInstallPpi (&mDelayedDispatchDesc);
+        ASSERT_EFI_ERROR (Status);
+        Status = PeiServicesNotifyPpi (&mDelayedDispatchNotifyDesc);
+        ASSERT_EFI_ERROR (Status);
+      }
+    }                                                                    // MS_CHANGE End
```
Comment (os-d) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: While you are making changes, can you change the MS_CHANGEs to MU_CHANGEs? Also, the end tag should go on the next brace down.

File: MdeModulePkg/Core/Pei/PeiMain.h [hunk 6]
```diff
@@ -316,7 +343,6 @@ struct _PEI_CORE_INSTANCE {
 ///
 #define PEI_CORE_INSTANCE_FROM_PS_THIS(a) \
   CR(a, PEI_CORE_INSTANCE, Ps, PEI_CORE_HANDLE_SIGNATURE)
-
```
Comment (os-d) [inline MdeModulePkg/Core/Pei/PeiMain.h]: This can be removed.

File: MdePkg/MdePkg.dec [hunk 7]
```diff
@@ -903,6 +903,25 @@
   # GUID used to provide initrd to linux via LoadFile2 protocol
```
Comment (os-d) [inline MdePkg/MdePkg.dec]: Can you also split this commit per pkg?

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 8]
```diff
@@ -1660,13 +1988,16 @@ PeiDispatcher (
...
                   PeimEntryPoint = (EFI_PEIM_ENTRY_POINT2)(UINTN)EntryPoint;
                 }
               }
 
               ASSERT (PeimEntryPoint != NULL);
+              // PERF_ENTRY_POINT_BEGIN (PeimFileHandle); // MU_CHANGE
               PeimEntryPoint (PeimFileHandle, (const EFI_PEI_SERVICES **)PeiServices);
-              // PERF_END (PeiServices, L"PEIM", PeimFileHandle, 0);
+              // PERF_END (PeiServices, L"PEIM", PeimFileHandle, 0); // MU_CHANGE
+              // PERF_ENTRY_POINT_END (PeimFileHandle); // MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: We can drop these commented additions?

File: MdeModulePkg/Library/UefiBootManagerLib/InternalBm.h [hunk 9]
```diff
@@ -68,7 +68,6 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/DxeServicesLib.h>
 #include <Library/ReportStatusCodeLib.h>
 #include <Library/CapsuleLib.h>
-#include <Library/PerformanceLib.h>
```
Comment (os-d) [inline MdeModulePkg/Library/UefiBootManagerLib/InternalBm.h:71]: Can we do the canonical style of commenting out with a MU_CHANGE tag?

File: MdeModulePkg/Library/UefiBootManagerLib/UefiBootManagerLib.inf [hunk 10]
```diff
@@ -59,7 +59,6 @@
   MemoryAllocationLib
   DxeServicesLib
   ReportStatusCodeLib
-  PerformanceLib
```
Comment (os-d) [inline MdeModulePkg/Library/UefiBootManagerLib/UefiBootManagerLib.inf:62]: Can we do the canonical style of commenting out with a MU_CHANGE tag?

**Review by os-d — APPROVED** [review]

File: MdePkg/Include/Ppi/DelayedDispatch.h (hunk 3 above)
Comment (Flickdm) [reply MdePkg/Include/Ppi/DelayedDispatch.h]: Accurate - having two copies causes the build to fail.
I could split this into it's own smaller PR. The problem is that Edk2 has this in MdePkg and we have it in MdeModulePkg
